src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -1067,34 +1124,13 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
                         if (addRangeMethod is null)
                             return false;
 
-                        if (spreadElement.EnumeratorInfoOpt is { } enumeratorInfo)
-                        {
-                            var iCollectionOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_ICollection_T);
-                            var iCollectionOfElementType = iCollectionOfTType.Construct(enumeratorInfo.ElementType);
-                            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
-
-                            // If collection has a struct enumerator but doesn't implement ICollection<T>
-                            // then manual `foreach` is always more efficient then using `AddRange` method
-                            if (enumeratorInfo.GetEnumeratorInfo.Method.ReturnType.IsValueType &&
-                                !enumeratorInfo.CollectionType.ImplementsInterface(iCollectionOfElementType, ref discardedUseSiteInfo))
-                            {
-                                return false;
-                            }
-                        }
-
-                        var type = rewrittenSpreadOperand.Type!;
-
-                        var useSiteInfo = GetNewCompoundUseSiteInfo();
-                        var conversion = _compilation.Conversions.ClassifyConversionFromType(type, addRangeMethod.Parameters[0].Type, isChecked: false, ref useSiteInfo);
-                        _diagnostics.Add(rewrittenSpreadOperand.Syntax, useSiteInfo);
-                        if (conversion.IsIdentity || (conversion.IsImplicit && conversion.IsReference))
+                        if (!ShouldUseAddRangeOrToListMethod(rewrittenSpreadOperand.Type!, spreadElement.EnumeratorInfoOpt?.GetEnumeratorInfo.Method, elementType))
```

Comment:
>rewrittenSpreadOperand.Type! [](http://example.com/codeflow?start=61&length=28)

Let's add a `Debug.Assert()` to the beginning of this lambda body (before the `if (addRangeMethod is null) return false;`), rather than using `!`.